### PR TITLE
Temporal indexing IV - range scans

### PR DESCRIPTION
**Pull request by fickludd**
Enable range scan functionality of all temporal index parts. 

~~Still missing test for some corner-cases!~~

~~Builds on other stuff, first commit is `Refactor range predicates to share base class`.~~

### CODE REVIEW

**Comment by fickludd** [comment]
I know this is conflicting, but hoping to merge the underlying PR and then rebase on master. Still review-able while waiting.

File: community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableTransactionState.java [hunk 1]
```diff
@@ -145,14 +146,14 @@ PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByNumber( IndexDescriptor
                                                              Number lower, boolean includeLower,
                                                              Number upper, boolean includeUpper );
 
-    PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByGeometry( IndexDescriptor index,
-                                                             PointValue lower, boolean includeLower,
-                                                             PointValue upper, boolean includeUpper );
-
     PrimitiveLongReadableDiffSets indexUpdatesForRangeSeekByString( IndexDescriptor index,
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableTransactionState.java]: Should this method also be removed?

File: community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java [hunk 2]
```diff
@@ -92,14 +92,43 @@ public static RangePredicate range( int propertyKeyId,
...
     }
 
+    /**
+     * Create IndexQuery for retrieving all indexed entries of the given value group.
+     */
+    public static RangePredicate range( int propertyKeyId, ValueGroup valueGroup )
+    {
+        if ( valueGroup == ValueGroup.GEOMETRY )
+        {
+            throw new IllegalArgumentException( "Cannot create GeometryRangePredicate without at specified CRS" );
```
Comment (tinwelint) [inline community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java]: `at` --> `a` ?

File: community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java [hunk 3]
```diff
@@ -377,7 +402,10 @@ public boolean acceptsValue( Value value )
             if ( value instanceof PointValue )
             {
                 PointValue point = (PointValue) value;
-                return point.withinRange( from, fromInclusive, to, toInclusive );
+                if ( point.getCRS() == crs )
```
Comment (tinwelint) [inline community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java]: `==` assuming too much? Better with `.equals`?

**Review by tinwelint — CHANGES_REQUESTED** [review]
Nice simplifications with the range query refactoring!

Had a few comments. And on the testing side, what are the corner cases that are missing, and if known why are you not adding those in this PR?

File: community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableTransactionState.java (hunk 1 above)
Comment (fickludd) [reply community/kernel/src/main/java/org/neo4j/storageengine/api/txstate/ReadableTransactionState.java]: Yes! Just forgot about it.

**Comment by fickludd** [comment]
@tinwelint I put in the corner-case note in the description when I first created this PR, which was mostly for building. We since explored my suspected corner-cases, so that's just old stale suspicion.

File: community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java [hunk 4]
```diff
@@ -402,19 +399,20 @@ public boolean acceptsValue( Value value )
             {
                 return false;
             }
-            //TODO Deal with other Geometries
             if ( value instanceof PointValue )
             {
                 PointValue point = (PointValue) value;
-                return point.withinRange( from, fromInclusive, to, toInclusive );
+                if ( point.getCRS().equals( crs ) )
```
Comment (OliviaYtterbrink) [inline community/kernel-api/src/main/java/org/neo4j/internal/kernel/api/IndexQuery.java:405]: This check is also done in `PointValue.withinRange`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java [hunk 5]
```diff
@@ -85,34 +84,39 @@ public PrimitiveLongResourceIterator query( IndexQuery... predicates ) throws In
...
+            case DATE:
+            case LOCAL_DATE_TIME:
+            case ZONED_DATE_TIME:
+            case LOCAL_TIME:
+            case ZONED_TIME:
+            case DURATION:
+                return instances[TEMPORAL].query( predicates );
+            default: // fall through
+            }
+            // TODO: support temporal range queries
```
Comment (OliviaYtterbrink) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:117]: this todo is outdated...

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java [hunk 6]
```diff
@@ -136,38 +140,37 @@ public void query( IndexProgressor.NodeValueClient cursor, IndexOrder indexOrder
...
+                return;
+            case GEOMETRY:
+                instances[SPATIAL].query( cursor, indexOrder, predicates );
+                return;
+            case TEXT:
+                instances[STRING].query( cursor, indexOrder, predicates );
+                return;
+            default: // fall through
+            }
+            // TODO: support temporal range queries
```
Comment (OliviaYtterbrink) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:171]: forgot to add support on this path...

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java [hunk 7]
```diff
@@ -204,16 +207,13 @@ public boolean hasFullValuePrecision( IndexQuery... predicates )
             Value value = ((ExactPredicate) predicate).value();
             return selector.select( instances, value ).hasFullValuePrecision( predicates );
         }
-        if ( predicate instanceof NumberRangePredicate )
+
+        if ( predicate instanceof RangePredicate && predicate.valueGroup() == ValueGroup.NUMBER )
         {
             return instances[NUMBER].hasFullValuePrecision( predicates );
         }
+        // TODO: support temporal range queries
```
Comment (OliviaYtterbrink) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:215]: this also missed being fixed...

File: community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Read.java [hunk 8]
```diff
@@ -111,9 +113,9 @@ public final void nodeIndexSeek(
             {
                 switch ( q.type() )
                 {
-                case rangeGeometric:
-                case rangeNumeric:
-                    if ( !reader.hasFullValuePrecision( q ) )
+                case range:
+                    ValueGroup valueGroup = q.valueGroup();
```
Comment (OliviaYtterbrink) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Read.java:117]: do we really need to check the value group here, isn't it enough to check the `hasFullValuePrecision`?

**Review by OliviaYtterbrink — APPROVED** [review]
Just some minor points, but most of them are already fixed in the next PR
